cmd/package-operator-manager/teardown/teardown.go

```diff
@@ -0,0 +1,185 @@
+			return fmt.Errorf("wait for crd to be gone: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// removePKOClusterPackageFinalizer removes the teardown finalizer from the PKO package.
+//
+// This allows it to be deleted when PKO managed CRDs are removed.
+func (t Teardown) removePKOClusterPackageFinalizer(ctx context.Context) error {
+	var pkg pkocore.ClusterPackage
+
+	if err := t.c.Get(ctx, types.NamespacedName{Name: bootstrap.ClusterPackageName}, &pkg); err != nil {
+		return fmt.Errorf("get PKO cluster package: %w", err)
+	}
+
+	if err := controllers.RemoveFinalizer(ctx, t.c, &pkg, bootstrap.ClusterPackageFinalizer); err != nil {
+		return fmt.Errorf("remove finalizer from PKO cluster package: %w", err)
+	}
+
+	return nil
+}
+
+// removeClusterObjectSetFinalizer removes cache finalizer from ClusterObjectSets.
+//
+// PKO doesn't clean them up properly so we do that here.
+func (t Teardown) removeClusterObjectSetFinalizer(ctx context.Context) error {
+	cobsl := pkocore.ClusterObjectSetList{}
+	if err := t.c.List(ctx, &cobsl); err != nil {
+		return fmt.Errorf("list PKO clusterobjectsets: %w", err)
+	}
+
+	for i := range cobsl.Items {
+		logr.FromContextOrDiscard(ctx).Info("remove cached finalizer from ClusterObjectSet", "name", &cobsl.Items[i].Name)
+
+		if err := controllers.RemoveFinalizer(ctx, t.c, &cobsl.Items[i], "package-operator.run/cached"); err != nil {
+			return fmt.Errorf("remove cached finalizer from clusterobjectset: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// removeCRDs removes CRDs that were installed with PKO and waits for them to be gone.
+func (t Teardown) removeCRDs(ctx context.Context) error {
+	selector, err := labels.ValidatedSelectorFromSet(map[string]string{"package-operator.run/package": "package-operator"})
+	if err != nil {
+		panic(err)
+	}
+
+	crdl := ext.CustomResourceDefinitionList{}
+
+	if err := t.c.List(ctx, &crdl, &client.ListOptions{LabelSelector: selector}); err != nil {
+		return fmt.Errorf("list crds: %w", err)
+	}
+
+	for i := range crdl.Items {
+		crd := &crdl.Items[i]
+
+		logr.FromContextOrDiscard(ctx).Info("deleting crd", "name", crd.Name)
+
+		if err := t.c.Delete(ctx, crd); err != nil {
+			return fmt.Errorf("delete crd: %w", err)
+		}
+	}
+
+	for i := range crdl.Items {
+		if err := wait.PollUntilContextCancel(ctx, 1*time.Second, true, utils.ConditionFnNotFound(t.c, &crdl.Items[i])); err != nil {
+			return fmt.Errorf("wait for crd to be gone: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// removePerms removes the ClusterRole and ClusterRoleBinding of PKO.
+func (t Teardown) removePerms(ctx context.Context) error {
+	if err := t.c.Delete(ctx, &rbac.ClusterRole{ObjectMeta: meta.ObjectMeta{Name: "package-operator-remote-phase-manager"}}); err != nil {
+		return fmt.Errorf("delete PKO crb: %w", err)
+	}
+
+	if err := t.c.Delete(ctx, &rbac.ClusterRoleBinding{ObjectMeta: meta.ObjectMeta{Name: "package-operator"}}); err != nil {
+		return fmt.Errorf("delete PKO crb: %w", err)
+	}
```

Comment:
Who cleans up the corresponding (Cluster)Roles?
Also: Who cleans up the pko deployment, serviceaccount and namespace? Does pko do that itself?

Reply:
> Who cleans up the corresponding (Cluster)Roles?
line 176 does :D
> Also: Who cleans up the pko deployment, serviceaccount and namespace? Does pko do that itself?
gets cleaned up by ownerref

Reply:
Oh well... advantage to those who can read :D